pantheon-bundle/src/main/java/com/redhat/pantheon/model/module/ModuleLocale.java

```diff
@@ -25,6 +28,61 @@ default ModuleVersion getOrCreateVersion(String name) {
         return getOrCreateChild(name, ModuleVersion.class);
     }
 
+    /**
+     * Archives the latest released version. Renames the version to the next available archive name.
+     * @return The newly archived {@link ModuleVersion}
+     */
+    default ModuleVersion archiveReleasedVersion() {
+        if(released().get() == null) {
+            // nothing to do
+            return null;
+        }
+
+        // Rename 'released' to the next sequence number
+        String archivedVersionName = generateNextArchiveVersionName();
+        try {
+            return SlingModels.rename(released().get(), archivedVersionName, ModuleVersion.class);
+        } catch (RepositoryException e) {
+            throw new RuntimeException("Problem archiving released version to '" + archivedVersionName + "'", e);
+        }
+    }
+
+    default ModuleVersion releaseDraftVersion() {
+        if(draft().get() == null) {
+            // nothing to do
+            return null;
+        }
+
+        // Rename 'draft' to 'released'
+        try {
+            // TODO externalize this magic string
+            return SlingModels.rename(draft().get(), "released", ModuleVersion.class);
+        } catch (RepositoryException e) {
+            throw new RuntimeException("Problem releasing draft version at '" + draft().get().getPath() + "'", e);
+        }
+    }
+
+    default ModuleVersion rollbackReleasedVersion() {
+        if(released().get() == null) {
+            // nothing to do
+            return null;
+        }
+
+        try {
+            // If there is no 'draft' version already, rename 'released' to 'draft'
+            if(draft().get() == null) {
+                // TODO externalize this magic string
+                return SlingModels.rename(released().get(), "draft", ModuleVersion.class);
```

Comment:
This would need to get a little bit more robust once we have assemblies - what if an assembly is referring to this module version, we can't simply turn it into a draft that might get overwritten with another upload... but, can you even unpublish a module that an assembly is currently depending on? Oof - makes my head hurt. This is fine for now.

Reply:
You are right, we will need to revisit this. All I was trying to do here was replicate whatever functionality the system offers before the change.